File: sdk/typescript/src/context.ts
/**
 * Type-safe context for request-scoped data.
 *
 * Uses branded types and TypeScript's structural typing for type safety.
 */

/**
 * Brand type for nominal typing.
 */
declare const brand: unique symbol;
type Brand<T, B> = T & { readonly [brand]: B };

/**
 * Context key type - branded for type safety.
 */
export type ContextKey<T> = Brand<symbol, T>;

/**
 * Creates a typed context key.
 */
export function contextKey<T>(description?: string): ContextKey<T> {
  return Symbol(description) as ContextKey<T>;
}

/**
 * Type-safe context storage.
 */
export class TypedContext {
  private readonly data = new Map<symbol, unknown>();
  private readonly headers = new Map<string, string>();

  /**
   * Sets a typed value.
   */
  set<T>(key: ContextKey<T>, value: T): this {
    this.data.set(key, value);
    return this;
  }

  /**
   * Gets a typed value.
   */
  get<T>(key: ContextKey<T>): T | undefined {
    return this.data.get(key) as T | undefined;
  }

  /**
   * Gets a typed value or throws.
   */
  require<T>(key: ContextKey<T>): T {
    const value = this.get(key);
    if (value === undefined) {
      throw new Error(`Required context key not found: ${key.toString()}`);
    }
    return value;
  }

  /**
   * Checks if a key exists.
   */
  has<T>(key: ContextKey<T>): boolean {
    return this.data.has(key);
  }

  /**
   * Deletes a value.
   */
  delete<T>(key: ContextKey<T>): boolean {
    return this.data.delete(key);
  }

  /**
   * Sets a header.
   */
  setHeader(name: string, value: string): this {
    this.headers.set(name.toLowerCase(), value);
    return this;
  }

  /**
   * Gets a header.
   */
  getHeader(name: string): string | undefined {
    return this.headers.get(name.toLowerCase());
  }

  /**
   * Gets all headers.
   */
  getHeaders(): ReadonlyMap<string, string> {
    return this.headers;
  }

  /**
   * Creates a child context with inherited values.
   */
  child(): TypedContext {
    const child = new TypedContext();
    for (const [key, value] of this.data) {
      child.data.set(key, value);
    }
    for (const [key, value] of this.headers) {
      child.headers.set(key, value);
    }
    return child;
  }
}

/**
 * Creates a new typed context.
 */
export function createContext(): TypedContext {
  return new TypedContext();
}

// Common context keys
export const CurrentUserId = contextKey<string>("CurrentUserId");
export const UserRoles = contextKey<readonly string[]>("UserRoles");
export const RequestId = contextKey<string>("RequestId");
export const RequestStartTime = contextKey<number>("RequestStartTime");

/**
 * User roles helper.
 */
export interface RolesHelper {
  readonly roles: readonly string[];
  has(role: string): boolean;
  hasAny(...roles: string[]): boolean;
  hasAll(...roles: string[]): boolean;
}

/**
 * Creates a roles helper.
 */
export function createRolesHelper(roles: readonly string[]): RolesHelper {
  return {
    roles,
    has: (role) => roles.includes(role),
    hasAny: (...checkRoles) => checkRoles.some((r) => roles.includes(r)),
    hasAll: (...checkRoles) => checkRoles.every((r) => roles.includes(r)),
  };
}

/**
 * Context builder for fluent API.
 */
export class ContextBuilder {
  private readonly ctx = new TypedContext();

  with<T>(key: ContextKey<T>, value: T): this {
    this.ctx.set(key, value);
    return this;
  }

  withHeader(name: string, value: string): this {
    this.ctx.setHeader(name, value);
    return this;
  }

  withUserId(userId: string): this {
    return this.with(CurrentUserId, userId);
  }

  withRoles(roles: readonly string[]): this {
    return this.with(UserRoles, roles);
  }

  withRequestId(requestId?: string): this {
    const id = requestId ?? `req_${Date.now().toString(36)}`;
    return this.with(RequestId, id);
  }

  build(): TypedContext {
    return this.ctx;
  }
}

/**
 * Creates a context builder.
 */
export function buildContext(): ContextBuilder {
  return new ContextBuilder();
}

File: sdk/go/context.go
package sdk

import (
	"context"
	"net/http"
	"sync"
)

// ContextKey is a typed key for context values.
type ContextKey[T any] struct {
	name string
}

// NewContextKey creates a new typed context key.
func NewContextKey[T any](name string) ContextKey[T] {
	return ContextKey[T]{name: name}
}

// String returns the key name.
func (k ContextKey[T]) String() string {
	return k.name
}

// Set stores a value in the context.
func (k ContextKey[T]) Set(ctx context.Context, value T) context.Context {
	return context.WithValue(ctx, k, value)
}

// Get retrieves a value from the context.
func (k ContextKey[T]) Get(ctx context.Context) (T, bool) {
	value, ok := ctx.Value(k).(T)
	return value, ok
}

// MustGet retrieves a value or panics.
func (k ContextKey[T]) MustGet(ctx context.Context) T {
	value, ok := k.Get(ctx)
	if !ok {
		panic("required context key not found: " + k.name)
	}
	return value
}

// GetOrDefault retrieves a value or returns a default.
func (k ContextKey[T]) GetOrDefault(ctx context.Context, defaultValue T) T {
	value, ok := k.Get(ctx)
	if !ok {
		return defaultValue
	}
	return value
}

// Common context keys
var (
	CurrentUserID   = NewContextKey[string]("CurrentUserID")
	UserRoles       = NewContextKey[[]string]("UserRoles")
	RequestID       = NewContextKey[string]("RequestID")
	RequestHeaders  = NewContextKey[http.Header]("RequestHeaders")
)

// RolesHelper provides role checking utilities.
type RolesHelper struct {
	roles []string
}

// NewRolesHelper creates a new roles helper.
func NewRolesHelper(roles []string) *RolesHelper {
	return &RolesHelper{roles: roles}
}

// Has checks if a role exists.
func (h *RolesHelper) Has(role string) bool {
	for _, r := range h.roles {
		if r == role {
			return true
		}
	}
	return false
}

// HasAny checks if any of the roles exist.
func (h *RolesHelper) HasAny(roles ...string) bool {
	for _, role := range roles {
		if h.Has(role) {
			return true
		}
	}
	return false
}

// HasAll checks if all roles exist.
func (h *RolesHelper) HasAll(roles ...string) bool {
	for _, role := range roles {
		if !h.Has(role) {
			return false
		}
	}
	return true
}

// Roles returns all roles.
func (h *RolesHelper) Roles() []string {
	return h.roles
}

// GetRolesHelper extracts roles from context and creates a helper.
func GetRolesHelper(ctx context.Context) *RolesHelper {
	roles, ok := UserRoles.Get(ctx)
	if !ok {
		return NewRolesHelper(nil)
	}
	return NewRolesHelper(roles)
}

// ContextBuilder builds a context with fluent API.
type ContextBuilder struct {
	ctx context.Context
}

// NewContextBuilder creates a new context builder.
func NewContextBuilder(ctx context.Context) *ContextBuilder {
	if ctx == nil {
		ctx = context.Background()
	}
	return &ContextBuilder{ctx: ctx}
}

// With adds a typed value to the context.
func With[T any](b *ContextBuilder, key ContextKey[T], value T) *ContextBuilder {
	b.ctx = key.Set(b.ctx, value)
	return b
}

// WithUserID adds a user ID to the context.
func (b *ContextBuilder) WithUserID(userID string) *ContextBuilder {
	b.ctx = CurrentUserID.Set(b.ctx, userID)
	return b
}

// WithRoles adds roles to the context.
func (b *ContextBuilder) WithRoles(roles []string) *ContextBuilder {
	b.ctx = UserRoles.Set(b.ctx, roles)
	return b
}

// WithRequestID adds a request ID to the context.
func (b *ContextBuilder) WithRequestID(requestID string) *ContextBuilder {
	b.ctx = RequestID.Set(b.ctx, requestID)
	return b
}

// Build returns the built context.
func (b *ContextBuilder) Build() context.Context {
	return b.ctx
}

// TypedContext wraps a context with additional typed storage.
type TypedContext struct {
	context.Context
	mu   sync.RWMutex
	data map[any]any
}

// NewTypedContext creates a new typed context.
func NewTypedContext(ctx context.Context) *TypedContext {
	if ctx == nil {
		ctx = context.Background()
	}
	return &TypedContext{
		Context: ctx,
		data:    make(map[any]any),
	}
}

// Set stores a typed value.
func (c *TypedContext) Set(key, value any) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.data[key] = value
}

// Get retrieves a typed value.
func (c *TypedContext) Get(key any) (any, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	value, ok := c.data[key]
	return value, ok
}

// GetTyped retrieves a typed value with type assertion.
func GetTyped[T any](c *TypedContext, key any) (T, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	value, ok := c.data[key]
	if !ok {
		var zero T
		return zero, false
	}
	typed, ok := value.(T)
	return typed, ok
}

// SetTyped stores a typed value with type safety.
func SetTyped[T any](c *TypedContext, key ContextKey[T], value T) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.data[key] = value
}
